File: fuentes/web/src-ts/objeto-visu.ts

import { Log }                  
from "./utilidades.js"
import { Vec3, Mat4 }
from "./vec-mat.js"

import { Cauce }
from "./cauce.js"

import { Textura }
from "./texturas.js" 

import { Material }
from "./material.js"

import { AplicacionWeb }
from "./aplicacion-web.js"
import { CauceBase } from "./cauce-base.js"

export abstract class ObjetoVisualizable
{
    
    /**
     * Color del objeto, null si no tiene
     */
    private color : Vec3 | null = null 

    public get tieneColor() : Boolean
    {
        return this.color != null
    }
    public get leerColor() : Vec3
    {
        if ( this.color == null )
            throw new Error(`intento de leer el color de un ObjetoVisualizable que no lo tiene (${this.nombre})`)
        return this.color
    }
    public set fijarColor( nuevo_color : Vec3 | null ) 
    {
        this.color = nuevo_color
    }

    /**
     * Textura del objeto:
     *   * undefined    --> hereda la textura que hubiera en el cauce cada vez que se visualiza
     *   * null         --> se visualiza sin textura    
     *   * en otro caso --> apunta a una textura que se usa para visualizar
     */
    private textura_act : Textura | null | undefined = undefined

    /**
     * Devuelve true si la textura no se hereda.
     */
    public get tieneTexturaDefinida() : Boolean 
    {
        if ( this.textura_act === undefined )
            return false 
        return true 
    }
    /**
     * Devuelve la textura del objeto (solo si la tiene definida)
     *   * Puede devolver 'null' (el objeto se dibuja sin textura)
     *   * O bien devolver un puntero a una textura (se dibuja con esa textura)
     */
    public get textura() : Textura | null 
    {
        if ( this.textura_act === undefined ) 
            throw new Error(`intento de leer la textura de un objeto que no tiene` )

        return this.textura_act
    }
    /**
     * Fija la textura del objeto.
     */
    public set textura( nueva_textura : Textura | null | undefined )
    {
        this.textura_act = nueva_textura
    } 


    /**
     * Matriz de modelado de este objeto (si no es nula),
     *   * Si está presente, es adicional a la que haya establecida en el cauce
     * (se hace pushMM antes de visualizar y popMM después)
     */
    private matrizm : Mat4 | null = null

    /**
     * Devuelve 'true' si la matriz de modelado no es nula
     */
    public get tieneMatrizModelado() : boolean 
    {
        return this.matrizm != null 
    }
    /**
     * Devuelve la matriz de modelado (no nula).
     * Si es nula produce un error
     */
    public get matrizModelado() : Mat4 
    { 
        if ( this.matrizm == null )  
            throw new Error("Intento de leer una matriz de modelado nula")
        return this.matrizm
    }
    /**
     * Fija la matriz de modelado (clona la que se le pasa como parámetro)
     */
    public set matrizModelado( nueva_matrizm : Mat4 | null )
    {
        if ( nueva_matrizm == null )
            this.matrizm = null 
        else 
            this.matrizm = nueva_matrizm.clonar()
    }

    /**
     * Si el objeto tiene definida su propia matriz de modelado
     * Guarda (push) la matriz de modelado en el cauce, y la compone con la actual
     * @param cauce 
     */
    public pushCompMM( cauce : Cauce )
    {
        if ( this.tieneMatrizModelado )
        {
            cauce.pushMM()
            cauce.compMM( this.matrizModelado )
        }
    }
    public popMM( cauce : Cauce )
    {
        if ( this.tieneMatrizModelado )
            cauce.popMM()
    }

    /**
     * Material del objeto (null si no tiene ninguno definido)
     */
    private material_act : Material | null = null 

    /**
     * Devuelve true si el objeto tiene un material propio, false en otro caso
     */
    public get tieneMaterial() : boolean 
    {
        return this.material_act != null 
    }
    /**
     * Si tiene material, lo devuelve, en otro caso se produce un error
     */
    public get material() : Material 
    {
        const nombref = "ObjetoVisualizable.material (getter)"
        if ( this.material_act == null )
            throw new Error(`${nombref}: intento de leer un material nulo.`)
        return this.material_act
    }

    /**
     * Cambia el material actual, se puede borrar (con null) o
     * bien poner uno nuevo (!= null)
     */
    public set material( nuevo_material : Material | null ) 
    {
        this.material_act = nuevo_material
    }


    /**
     * Nombre actual del objeto
     */
    private nombre_act  : string = "no asignado" 

    /**
     * Devuelve el nombre del objeto 
     */
    public get nombre() : string
    {
        return this.nombre_act 
    }
    /**
     * Cambia el nombre del objeto.
     */
    public set nombre( nuevo_nombre : string ) 
    {
        this.nombre_act = nuevo_nombre
    }

    /**
     * Parámetro 'S', usado para gestionar diversos aspectos del objeto
     */
    private param_S_act : number = AplicacionWeb.valor_inicial_param_S

    /**
     * Fija el nuevo valor del parámetro S
     */
    public set param_S( nuevo_param_s : number )  
    {
        this.param_S_act = nuevo_param_s
    }

    public get param_S() : number 
    {
        return this.param_S_act
    }

    /**
     * Visualiza el objeto. este método debe ser redefinido en clases derivadas
     * El objeto se visualiza usando el cauce definido en la aplicación (clase 'Cauce')
     */
    public abstract visualizar() : void ;


    /**
     * Visualiza el objeto sobre un cauce básico, únicamente la geometría, nada más
     * (se supone que el cauce está activo al llamar a este método)
     */
    public visualizarGeometria( cauceb : CauceBase ) : void 
    {
        const nombref : string = `ObjetoVisualizable.visualizarGeometria (${this.nombre}):`
        Log(`${nombref}: advertencia: no se hace nada: este objeto es de una clase que no redefine el método`)
    }

    /**
     * Visualiza las aristas del objeto. Este método puede ser redefinido en clases derivadas, si 
     * no se hace, el método no hace nada (eso implica que ese objeto no tiene aristas que se pueden visualizar 
     * o que no se ha implementado esto)
     */
    public visualizarAristas() : void  
    {
        Log(`ObjetoVisible.visualizarAristas: no se hace nada: este objeto es de una clase que no redefine el método`)
    }
    

    /**
     * Visualiza las normales del objeto. Este método puede ser redefinido en clases derivadas, si 
     * no se hace, el método no hace nada (eso implica que ese objeto no tiene normales que se pueden visualizar 
     * o que no se ha implementado esto)
     */
    public visualizarNormales() : void 
    {
        Log(`ObjetoVisible.visualizarNormales: no se hace nada: este objeto es de una clase que no redefine el método`)
    }
    
    /**
     * Guarda el estado actual de los (algunos/todos?) los uniforms y 
     * lo actualiza según este objeto
     * @param cauce cauce sobre el que se fija y que guarda el estado.
     */
    public guardarCambiarEstado( cauce : Cauce )
    {
        if ( this.tieneMaterial )
        {
            cauce.pushMaterial()
            cauce.fijarMaterial( this.material )
        }

        if ( this.tieneTexturaDefinida )
        {
            cauce.pushTextura()
            cauce.fijarTextura( this.textura )
        }

        if ( this.tieneColor )
        {
            cauce.pushColor()
            cauce.fijarColor( this.leerColor )
        }

        this.pushCompMM( cauce )
    }

    /**
     * Restaura el estado previo a la última llamada a 'guardarFijarEstado'
     * @param cauce cauce desde donde se restaura el estado.
     */
    public restaurarEstado( cauce: Cauce )
    {
        this.popMM( cauce )

        if ( this.tieneColor )
            cauce.popColor()   

        if ( this.tieneTexturaDefinida )
            cauce.popTextura()

        if ( this.tieneMaterial )
            cauce.popMaterial()
    }
}

File: fuentes/moviles/proyecto.android-studio/app/src/main/java/mds/pcg1/objeto-visu.kt
// *********************************************************************
// **
// ** Asignatura: PCG (Programación del Cauce Gráfico).
// **
// ** Clases relacionadas con objetos visualizables
// ** Copyright (C) 2024 Carlos Ureña
// **
// ** Clase: ObjetoVisualizable
// **
// ** This program is free software: you can redistribute it and/or modify
// ** it under the terms of the GNU General Public License as published by
// ** the Free Software Foundation, either version 3 of the License, or
// ** (at your option) any later version.
// **
// ** This program is distributed in the hope that it will be useful,
// ** but WITHOUT ANY WARRANTY; without even the implied warranty of
// ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// ** GNU General Public License for more details.
// **
// ** You should have received a copy of the GNU General Public License
// ** along with this program.  If not, see <http://www.gnu.org/licenses/>.
// **
// *********************************************************************

package mds.pcg1.objeto_visu

import android.util.Log
import mds.pcg1.aplicacion.AplicacionPCG
import mds.pcg1.cauce.CauceBase
import mds.pcg1.material.Material
import mds.pcg1.texturas.*
import mds.pcg1.vec_mat.*
import mds.pcg1.utilidades.*

// *************************************************************************************************

/**
 * Clase abstracta Objetos Visualizables:
 * Un objeto visualizable es un objeto que se puede visualizar invocando el método 'visualizar'
 *
 */
abstract class ObjetoVisualizable
{
    // ---------------------------------------------------------------------------------------------

    /**
     * nombre del objeto (es una String que lo identifica, por defecto no tiene)
     */
    var nombre : String = "no asignado" // nombre del objeto

        get() = field
        set( nuevo_nombre ) { field = nuevo_nombre }

    // ---------------------------------------------------------------------------------------------

    /**
     * Color específico del objeto (nulo si no tiene un color específico, en ese caso al
     * visualizarse en un cauce, se visualiza con el color que tenga el cauce en ese momento).
     * Por defecto es null.
     */
    var color : Vec3? = null // color del objeto, null si no tiene (inicialmente no tiene)

        get() = field
        set( nuevo_color ) { field = nuevo_color }

    // ---------------------------------------------------------------------------------------------

    /**
     * Textura del objeto (nulo si no tiene una textura, en ese caso se visualiza con la que haya
     * activa en el cauce, si hay alguna).
     */
    var textura : Textura? = null

        get() = field
        set( nueva_textura ) { field = nueva_textura }

    // ---------------------------------------------------------------------------------------------

    /**
     * Matriz de modelado de este objeto (si no es nula),
     * Si está presente, es adicional a la que haya establecida en el cauce
     * (se hace pushMM antes de visualizar y popMM después)
     */
    var matrizm : Mat4? = null

        get() = field
        set( nueva_matriz ) { field = nueva_matriz }

    // ---------------------------------------------------------------------------------------------

    /**
     * Material de este objeto (si no es nula),
     * Si está presente, el objeto se debe visualizar con este material
     * Si es nulo, el objeto se visualiza con el material que hubiera activo en el cauce
     */
    private var material : Material? = null

        get() = field
        set( nuevo_material ) { field = nuevo_material }

    // ---------------------------------------------------------------------------------------------
    /**
     * Visualiza el objeto. este método debe ser redefinido en clases derivadas
     */
    abstract fun visualizar()

    // ---------------------------------------------------------------------------------------------
    /**
     * Visualiza las aristas del objeto. Este método puede ser redefinido en clases derivadas, si
     * no se hace, el método no hace nada (eso implica que ese objeto no tiene aristas que se pueden visualizar
     * o que no se ha implementado esto)
     */
    open fun visualizarAristas()
    {
        Log.v( TAG, "El objeto '${nombre}' no tiene método para visualizar aristas ('visualizarAristas')" )
    }
    // ---------------------------------------------------------------------------------------------

    /**
     * Visualiza las normales del objeto. Este método puede ser redefinido en clases derivadas, si
     * no se hace, el método no hace nada (eso implica que ese objeto no tiene normales que se pueden visualizar
     * o que no se ha implementado esto)
     */
    open fun visualizarNormales()
    {
        Log.v( TAG, "El objeto '${nombre}' no tiene método para visualizar normales ('visualizarNormales')." )
    }
    // ---------------------------------------------------------------------------------------------

    /**
     * redefinir como true para objetos que tengan normales
     * (se usa iluminación solo si el objeto tiene normales)
     */
    open val tieneNormales : Boolean get()
    {
        return false
    }
    // ---------------------------------------------------------------------------------------------

    /**
     * guardar el estado de ciertos uniforms del cauce, y los cambia según este objeto
     *   - color
     *   - textura (que puede ser nula)
     *
     *   TODO: las texturas no se deberían guardar en el companion object, sino en una pila propia de cada cauce.
     */
    fun guardarCambiarEstado( cauce : CauceBase )
    {
        //val TAGF = "[${object {}.javaClass.enclosingMethod?.name?: nfnd}]"

        // si tiene matriz de modelado, guardar la actual y componerla en el cauce
        matrizm?.let {
            cauce.pushMM()
            cauce.compMM( it )
        }

        // si tiene color, guardar estado de color y cambiarlo
        color?.let {
            cauce.pushColor()
            cauce.fijarColor( it )
        }

        // guardar el estado de las texturas y cambiarlo
        // (la textura no se 'hereda' como el color, o bien el objeto tiene (textura!= null) o no tiene (textura == null)
        cauce.pushTextura()
        cauce.fijarTextura( textura )  // (desactiva texturas si textura == null)

        // si tiene material específico, lo activa, en otro caso se usa el que tenga el cauce
        material?.let {
            cauce.pushMaterial()
            cauce.fijarMaterial( it )
        }
    }
    // ---------------------------------------------------------------------------------------------
    /**
     * restaura el estado de ciertos uniforms del cauce (debe estar pareado con 'guardarEstado')
     */
    fun restaurarEstado( cauce : CauceBase )
    {
        //val TAGF = "[${object {}.javaClass.enclosingMethod?.name?: nfnd}]"

        // si tiene material, restaurar el color anterior
        if ( material != null )
            cauce.popMaterial()

        // restaurar el estado de las texturas
        cauce.popTextura( )

        // si tiene color, restaurar el estado de color
        if ( color != null )
            cauce.popColor()

        // si tiene matriz de modelado, restaurar la que había antes
        if ( matrizm  != null )
            cauce.popMM()
    }
}

// *************************************************************************************************